### PR TITLE
Restore auth when calling API only if needed

### PR DESCRIPTION
**Pull request by tsteur**
Also added tests, there were none before...

### CODE REVIEW

File: core/API/Request.php [hunk 1]
```diff
@@ -263,27 +292,53 @@ public static function reloadAuthUsingTokenAuth($request = null)
...
+         * Plugins that provide authentication capabilities should subscribe to this event
+         * and make sure the global authentication object (the object returned by `StaticContainer::get('Piwik\Auth')`)
+         * is setup to use `$token_auth` when its `authenticate()` method is executed.
+         *
+         * @param string $token_auth The value of the **token_auth** query parameter.
+         */
+        Piwik::postEvent('API.Request.authenticate', array($tokenAuth));
+        Access::getInstance()->reloadAccess();
+        SettingsServer::raiseMemoryLimitIfNecessary();
+    }
```
Comment (mnapoli) [inline core/API/Request.php]: It's maybe a matter of taste but I'm not a huge fan of `reloadAuthUsingTokenAuth()` actually calling `doReloadAuthUsingTokenAuth()`, it's a bit redundant and makes the code harder to navigate (eg. to illustrate what prevents from having another `doReallyReloadAuthUsingTokenAuth()`).

Maybe something like that?

```php
public static function reloadAuthUsingTokenAuth()
{
    if (! self::shouldReloadAuthUsingTokenAuth($request)) {
        return;
    }
    // do the thing
}
```

**Comment by mnapoli** [comment]
Except the minor comment :+1: Did you see a significant performance difference?

File: core/API/Request.php (hunk 1 above)
Comment (tsteur) [inline core/API/Request.php]: I can rename it to `doReallyReloadAuthUsingTokenAuth`. That method is called by `reloadAuthUsingTokenAuth` but also when restoring the token. Here we definitely want to reload auth even `shouldReloadAuthUsingTokenAuth` would return `false`. That's why I added that method.

Comment (tsteur) [inline core/API/Request.php]: Renamed it to `forceReloadAuthUsingTokenAuth`

Comment (mnapoli) [inline core/API/Request.php]: :+1: better choice
